File: pyUnrealLiveLink/examples/circling_transform.py

# 
# Copyright (c) 2025 Patrick Palmer
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#

import sys
import math
import time
import pyUnrealLiveLink as pyuell

# number of circles to perform 
CIRCLES = 30

# size in Unreal units of the circle on the X-Y plane 
CIRCLE_RADIUS = 200.0

# number of steps to perform the circles
STEP_COUNT = 10000

# Z axis location -- number of units above origin
HEIGHT = 100

angle = 0.0
world_time = 0.0
	
rc = pyuell.load();
if rc:
    print(f"error: unable to load Unreal Live Link shared object (error {rc})")
    sys.exit(1)

print("Starting...\n")

pyuell.set_provider_name("CirclingTransform")
pyuell.start_live_link()

# set up transform live link role
pyuell.set_transform_structure("circle", pyuell.Properties());
	
# initialize the per frame transform structure 
xform = pyuell.Transform()

# calculate the radian step
step = 2.0 * math.pi * float(CIRCLES) / float(STEP_COUNT)

# 24 fps
frame_time = 1.0 / 24.0

start_time = time.time()

# loop STEP_COUNT times sending data every 24 fps (not exact)
for i in range(STEP_COUNT):

    if (i % 96) == 0:
        print(f'frame {i}')

    x = math.sin(angle) * CIRCLE_RADIUS
    y = math.cos(angle) * CIRCLE_RADIUS
    z = HEIGHT
    xform.translation = (x, y, z)

    pyuell.update_transform_frame("circle", world_time, pyuell.Metadata(), pyuell.PropertyValues(), xform)
		
    angle += step
		
    # sleep 1 frame time
    time.sleep(frame_time)

    world_time += frame_time

end_time = time.time()
print(f"Done. Took {end_time - start_time} seconds.")

pyuell.unload()

sys.exit(0)

File: examples/CirclingTransform.c
/** 
 * Copyright (c) 2020 Patrick Palmer, The Jim Henson Company.
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include "UnrealLiveLinkCInterfaceAPI.h"
#ifdef WIN32
#include <windows.h>
#endif

#include <stdio.h>
#include <time.h>
#include <math.h>

/* number of circles to perform */
#define CIRCLES 30

/* size in Unreal units of the circle on the X-Y plane */
#define CIRCLE_RADIUS 200

/* number of steps to perform the circles */
#define STEP_COUNT 10000

/* Z axis location -- number of units above origin */
#define HEIGHT 100

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif


int main()
{
	int i;
	int rc;
	time_t t;
	double angle = 0.0;
	double worldTime = 0.0;
	
#ifdef WIN32
	const char * sharedObj = "UnrealLiveLinkCInterface.dll";
#else
	const char * sharedObj = "libUnrealLiveLinkCInterface.so";
	struct timespec ts;

	ts.tv_sec = 0;
	ts.tv_nsec = 16 * 1000000;
#endif

	rc = UnrealLiveLink_Load(sharedObj);
	if (rc != UNREAL_LIVE_LINK_OK)
	{
		printf("error: unable to load %s (error %d)\n", sharedObj, rc);
		return 1;
	}

	UnrealLiveLink_SetProviderName("CirclingTransform");
	rc = UnrealLiveLink_StartLiveLink();
	if (rc != UNREAL_LIVE_LINK_OK) 
	{
		printf("error: unable to start live link (error %d)\n", rc);
		return 1;
	}

	printf("Starting...\n");
	t = time(NULL);

	/* set up transform live link role */
	UnrealLiveLink_SetTransformStructure("circle", 0);
	
	/* initialize the per frame transform structure */
	struct UnrealLiveLink_Transform xform;
	UnrealLiveLink_InitTransform(&xform);
	xform.translation[2] = HEIGHT;

	/* calculate the radian step */
	const double step = 2.0 * M_PI * CIRCLES / STEP_COUNT;

	/* loop STEP_COUNT times sending data every 16ms (not exact) */
	for (i = 0; i < STEP_COUNT; i++)
	{
		xform.translation[0] = (float)sin(angle) * CIRCLE_RADIUS;
		xform.translation[1] = (float)cos(angle) * CIRCLE_RADIUS;

		UnrealLiveLink_UpdateTransformFrame("circle", worldTime, 0, 0, &xform);
		
		angle += step;
		
		/* sleep 16ms */
#ifdef WIN32
		Sleep(16);
#else
		nanosleep(&ts, &ts);
#endif

		worldTime += 16.0;
	}	

	printf("Done. Took %lld seconds.\n", time(NULL) - t);

	UnrealLiveLink_Unload();

	return 0;
}

